src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.StandardTextDocument.cs

```diff
@@ -242,6 +243,33 @@ public uint GetItemId()
                     ? itemId
                     : (uint)VSConstants.VSITEMID.Nil;
             }
+
+            /// <summary>
+            /// A wrapper for a <see cref="TextLoader"/> that ensures we are watching file contents prior to reading the file.
+            /// </summary>
+            private sealed class FileChangeTrackingTextLoader : TextLoader
+            {
+                private readonly FileChangeTracker _fileChangeTracker;
+                private readonly TextLoader _innerTextLoader;
+
+                public FileChangeTrackingTextLoader(FileChangeTracker fileChangeTracker, TextLoader innerTextLoader)
+                {
+                    _fileChangeTracker = fileChangeTracker;
+                    _innerTextLoader = innerTextLoader;
+                }
+
+                public override Task<TextAndVersion> LoadTextAndVersionAsync(Workspace workspace, DocumentId documentId, CancellationToken cancellationToken)
+                {
+                    _fileChangeTracker.EnsureSubscription();
```

Comment:
Does `EnsureSubscription` require a UI thread?  Is this only called from it?  Also, is a `WhenAll` between the subscription and the read enough?  I suppose you're worried about a race where we read, then a change happens and then the subscription task finishes?

Reply:
So many questions!

> Does EnsureSubscription require a UI thread?

No.

> Is this only called from it?

Unsure what 'this' and 'it' are in this context, but EnsureSubscription is also used for metadata reference and analyzer references too.

> Also, is a WhenAll between the subscription and the read enough? I suppose you're worried about a race where we read, then a change happens and then the subscription task finishes?

Yes, that wouldn't work for that reason. What *could* be done (and I'm planning on doing in my big PR) is for these cases we can just do the subscription asynchronously, and when it completes check if the timestamp on the disk a second time to see if it matches what was read when we did the initial read.

Reply:
'This' was the method, and 'it' was the UI thread.  The question was only relevant if the answer to the first was yes.